kyaml/kio/byteio_writer.go

```diff
@@ -125,7 +128,29 @@ func (w ByteWriter) Write(nodes []*yaml.RNode) error {
 	for i := range nodes {
 		items.Content = append(items.Content, nodes[i].YNode())
 	}
-	err := errors.Wrap(encoder.Encode(doc))
+	rNode := yaml.RNode{}
+	rNode.SetYNode(doc)
+	err := w.encode(encoder, &rNode)
 	yaml.UndoSerializationHacksOnNodes(nodes)
 	return err
 }
+
+// encode encodes the input RNode to appropriate node format based on file path extension
+func (w ByteWriter) encode(encoder *yaml.Encoder, node *yaml.RNode) error {
+	for _, g := range JSONMatch {
+		if match, err := filepath.Match(g, filepath.Ext(w.Path)); err != nil {
+			return errors.Wrap(err)
+		} else if match {
+			json, err := yaml.ConvertYamlNodeToJSONString(node)
```

Comment:
I'm not sure encoding each node as JSON individually works -- the whole collection needs to a be a json list right?  Or maybe the list needs to have exactly 1 item?  Yaml can contain multiple entries in a single byte stream without them being in a list, but JSON cannot.

This is going to be a bit more complicated to detect the right behavior.

Reply:
Probably error if there is more than 1 item and encoding is json

Reply:
Done. Added unit test to throw error if there are multiple json items.